### ISSUE TITLE
Less Hyphenation in Titles

### ISSUE THREAD

**Issue by tpreusse** [orbiting/pdf#128]
there is already logic punishing hyphenation when not justified:
https://github.com/diegomura/react-pdf/blob/3ab9b8d8d7ee9cb5cd49e5f2017a02c85be49968/src/layout/linebreaker/formatter.js#L70

However there is still too much.

Examples:
<img width="820" alt="screen shot 2018-10-25 at 16 50 50" src="https://user-images.githubusercontent.com/410211/47509632-ce14a480-d876-11e8-90f4-0926a82b7ce1.png">

<img width="825" alt="screen shot 2018-10-25 at 16 50 58" src="https://user-images.githubusercontent.com/410211/47509649-d371ef00-d876-11e8-95eb-0b9cd2e52b1f.png">

Maybe an option to disable for certain elements (e.g. titels) and only break at soft hyphen that are already in the string?

Basically in titles we don't want to break words unless absolutely necessary to fit in the page or where we manually placed soft hyphens.

**Comment by diegomura** [comment]
Makes sense! What about passing a prop down to `Text` about the hyphenation tolerance that text should have? That way we can leave that option to other users to customize that as they want. I'll have to see how easy that would be with the current setup, but I think I can make it work

**Comment by tpreusse** [comment]
Yeah something like that but not sure if customising penalty alone work. Maybe it need to be possible to even turn off Knuth and Plass for an `Text` and use a different, simpler algo which only breaks words exceeding the width? But maybe a super high penalty ultimately has the same effect.

**Comment by diegomura** [comment]
I don't think there's need to code a new strategy. The `formatter` solution that we already have on react-pdf already fallbacks to a solution similar to what you described.

Here's a quick test:
<img width="809" alt="screen shot 2018-10-26 at 12 05 49 pm" src="https://user-images.githubusercontent.com/5600341/47575067-8f82f680-d917-11e8-9d52-184baaa9b0ef.png">

And when word is too long to the available width:
<img width="809" alt="screen shot 2018-10-26 at 12 05 58 pm" src="https://user-images.githubusercontent.com/5600341/47575064-8c880600-d917-11e8-95ac-55a264f97274.png">

Unfortunately the fallbacks is applied after trying several times to apply K&P (that its how currently works). I'm going to try to decrease the amount of retries and see how it behaves in a bunch of docs. This will have a positive impact on performance too

**Comment by diegomura** [comment]
@tpreusse : https://github.com/orbiting/pdf/pull/130/files#diff-c3d0cf76cc4bddb491a943fcb5737e39R21

Thoughts?

I think this new api has two sides. On one hand it feels that we are exposing an certain complexity to the user (what is 600? and 10000?), but on the other hand we are providing full control on how hyphenation will work, making it more powerful.

For me that's acceptable, but maybe it worths a discussion.

**Comment by tpreusse** [comment]
I think it's acceptable. And pretty sweet advanced feature.

**Comment by diegomura** [comment]
Done for H2 as well.
I ran the test suite and the generation time decreased significantly. For 740 documents (slightly more than some days ago) locally the script ran in ~7:30 minutes. 
Didn't know that I had such a bottleneck in work wrapping! I'll visually test some documents to ensure everything is fine though
